### PR TITLE
CR6 OS

### PR DESCRIPTION
**Pull request by jacksonshanks**

### CODE REVIEW

**Review by rumachan — DISMISSED** [review]
Looks good with the date formats changed.

**Comment by elidana** [comment]
Hi @jacksonshanks , I think you always need to provide a starting date to each firmware entry, that might be why the build is failing 
entries such as 
`Campbell Scientific,CR6,4108,,,9999-01-01T00:00:00Z,wairakei` (and all other similar entries)
will fail as
- there is no start date
- there is no firwmare version associated with the equipment serial number (which is the aim of this firmware file).

For the firmware file, as `start date`, you can use a time before the equipment was deployed at a site, provided that it was not installed elsewhere.

If, instead, your idea was to record when a specific equipment (identified via its serial number) was installed at a specific location, you will need to use another file, not the `firmware.csv`

Feel free to give me a call, I might be missing something obvious here and my comment might not be relevant.

**Review by elidana — CHANGES_REQUESTED** [review]
Hi @jacksonshanks ,

this pull request will need to be modified to pass the automatic tests. 
I am not sure if this is a placeholder to start tracking firmware information of sensors installed at different sites but, for the info to go in the firmware file you will need to
- include start time and end time to each entry (start time should reflect the time when the firmware was operating at a particular sensor)
- remove the information about installation location (last entry on each line). This information should go in a different file, and the link between sensor serial number and station should go there (for Low Rate Data loggers I believe the file should be datalogger, but @rumachan can confirm that. 

Happy to have a chat about what you are trying to include in delta and find the best place to collect the information if you wish..

Thanks!

**Comment by jacksonshanks** [comment]
Hi @elidana. i was just talking with richard about this. we are happy to make this work however it is best for the delta system and repos. at the moment this unmerged file is a place holder for the information until we have collected it all.

@rumachan can you lead this conversation please, as we are filling this a request from your pervious ticket. its possible that tis file may not be the best place for all this? https://github.com/GeoNet/data-logger/issues/205#issuecomment-958102519

**Comment by elidana** [comment]
Hi @jacksonshanks , sounds good, thanks for clarifying that!

Feel free to use this as a placeholder until you need that in the current format, if that works for you.
We just won't be merging this pr for now , until all info are placed in the related table.

@rumachan we can have a chat about that and move from there if that works for you

**Review by rumachan — CHANGES_REQUESTED** [review]
As well as adding records with the new firmware version, which end at 9999-01-01, entries for the past firmware need to remain in the file, with a suitable end dates. 

There is an entry (line 9) with date in the incorrect format. This suggests the file you are editing is not current, as checks would fail.

**Comment by ricardojohansson** [comment]
@rumachan can I fill in the blanks (where logger OS or start date is unknown) by using something obviously incorrect?

eg
CR6.Std.00.00 for all entries where OS version is yet to be confirmed
1901-01-01 for all entries where start date is yet to be confirmed

**Review by rumachan — APPROVED** [review]

**Comment by rumachan** [comment]
Complaining  about 2006-01-02T15:04:05Z. That's not in your work, but somewhere in the file.

**Comment by rumachan** [comment]
I'm not seeing where this is failing

+go test ./tests
--- FAIL: TestConsistency (0.05s)
    consistency_test.go:18: unable to load list file ../install/firmware.csv: parsing time "" as "2006-01-02T15:04:05Z": cannot parse "" as "2006"
--- FAIL: TestFirmwareHistory (0.00s)
    consistency_test.go:18: unable to load list file ../install/firmware.csv: parsing time "" as "2006-01-02T15:04:05Z": cannot parse "" as "2006"
--- FAIL: TestFirmwareHistory_Assets (0.00s)
    consistency_test.go:18: unable to load list file ../install/firmware.csv: parsing time "" as "2006-01-02T15:04:05Z": cannot parse "" as "2006"
--- FAIL: TestFirmwareHistory_DeployedReceiver (0.00s)
    consistency_test.go:18: unable to load list file ../install/firmware.csv: parsing time "" as "2006-01-02T15:04:05Z": cannot parse "" as "2006"
FAIL
FAIL	github.com/GeoNet/delta/tests	0.491s
FAIL
++error_handler
++echo 'Trapped error - Unknown Error'
Trapped error - Unknown Error

**Comment by ricardojohansson** [comment]
@rumachan I just found this comment above

> Hi @jacksonshanks ,
> 
> this pull request will need to be modified to pass the automatic tests. I am not sure if this is a placeholder to start tracking firmware information of sensors installed at different sites but, for the info to go in the firmware file you will need to
> 
> * include start time and end time to each entry (start time should reflect the time when the firmware was operating at a particular sensor)
> * **remove the information about installation location (last entry on each line)**. This information should go in a different file, and the link between sensor serial number and station should go there (for Low Rate Data loggers I believe the file should be datalogger, but @rumachan can confirm that.
> 
> Happy to have a chat about what you are trying to include in delta and find the best place to collect the information if you wish..
> 
> Thanks!

**Comment by elidana** [comment]
Ditto rich! That comment was to try to capture the failures

@rumachan if you want to have a chat I'd be happy to discuss how we can potentially implement all those metadata into delta based on what we do for seismic and gnss sensors

**Comment by ricardojohansson** [comment]
So I have to delete everything after the last comma on each line?

**Comment by ricardojohansson** [comment]
@rumachan I checked (delta)dataloggers.csv but it currently just lists their serial numbers... Should I delete the location 'notes' from the firmware.csv and add this detail to dataloggers.csv, or are the loggers locations listed somewhere else?

**Comment by rumachan** [comment]
dataloggers.csv is just to establish what dataloggers exist, not exactly where they are, which is why there is no location in that file. The locations of the stations/sites are in network/stations and network/sites. Please remove any location information from firmware.csv as an instruments firmware is not associated with where an instrument is installed.

**Comment by ricardojohansson** [comment]
Humpf

File: install/firmware.csv [hunk 1]
```diff
@@ -2,6 +2,30 @@ Make,Model,Serial,Version,Start Date,End Date,Notes
 "Allen Osborne Associates, Inc.",ROGUE SNR-8000,132,3.2,1995-05-31T00:00:00Z,1999-08-12T23:10:00Z,
 "Allen Osborne Associates, Inc.",ROGUE SNR-8000,132,3.2.32.8,1999-08-12T23:10:01Z,9999-01-01T00:00:00Z,
 "Allen Osborne Associates, Inc.",ROGUE SNR-8000,133,3.2.32.8,1995-01-01T00:00:00Z,9999-01-01T00:00:00Z,
+Campbell Scientific,CR6,3066,CR6.Std.04,2018-08-10T00:00:00Z,2021-11-18T23:59:00Z,
+Campbell Scientific,CR6,3066,CR6.Std.11.01,2021-11-19T00:00:00Z,9999-01-01T00:00:00Z,
+Campbell Scientific,CR6,4108,CR6.Std.11.01,1901-01-01T00:00:00Z,9999-01-01T00:00:00Z,
+Campbell Scientific,CR6,4109,CR6.Std.09.02,2021-10-15T00:00:00Z,9999-01-01T00:00:00Z,
+Campbell Scientific,CR6,4110,CR6.Std.04,2018-08-28T00:00:00Z,2021-11-18T23:59:00Z,
+Campbell Scientific,CR6,4110,CR6.Std.11.01,2021-11-19T00:00:00Z,9999-01-01T00:00:00Z,
+Campbell Scientific,CR6,4111,CR6.Std.11.01,2021-17-11T14:10:00Z,9999-01-01T00:00:00Z,
```
Comment (elidana) [inline install/firmware.csv:11]: Hi @ricardojohansson , this entry has the wrong month (`17`), that is why the pr is failing

**Review by elidana — CHANGES_REQUESTED** [review]
a typo in a date `2021-17-...` that is making the pr to fail

File: install/firmware.csv (hunk 1 above)
Comment (ricardojohansson) [reply install/firmware.csv]: Thanks @elidana  - I've fixed that... but it's still failing

Comment (elidana) [reply install/firmware.csv]: Yes, I can see new and different error messages now popping up, but at least this is one less thing

**Comment by elidana** [comment]
### Errror n. 1 
include logger serial numbers in `assets/dataloggers.csv` (need to be in alphabetical and numerical order)
e.g.: 
```diff
 Campbell Scientific,CR6,4117,,
+Campbell Scientific,CR6,11300,,
 Campbell Scientific,CR6,15365,,
```


### Error n. 2
there is another error that states "identical lines" not sure where this one come from 
```
+ go test ./tests
--- FAIL: TestConsistency (0.15s)
    --- FAIL: TestConsistency/check_file_consistency:_firmware (0.01s)
        consistency_test.go:39: unexpected firmware.csv content -got/+exp
              strings.Join({
              	... // 2 identical lines
              	`"Allen Osborne Associates, Inc.",ROGUE SNR-8000,132,3.2.32.8,1999-08-12T23:10:01Z,9999-01-01T00:00:00Z,`,
              	`"Allen Osborne Associates, Inc.",ROGUE SNR-8000,133,3.2.32.8,1995-01-01T00:00:00Z,9999-01-01T00:00:00Z,`,
            - 	"Campbell Scientific,CR6,3066,CR6.Std.04,2018-08-10T00:00:00Z,2021-11-18T23:59:00Z,",
            - 	"Campbell Scientific,CR6,3066,CR6.Std.11.01,2021-11-19T00:00:00Z,9999-01-01T00:00:00Z,",
            - 	"Campbell Scientific,CR6,4108,CR6.Std.11.01,1901-01-01T00:00:00Z,9999-01-01T00:00:00Z,",
            - 	"Campbell Scientific,CR6,4109,CR6.Std.09.02,2021-10-15T00:00:00Z,9999-01-01T00:00:00Z,",
```

not sure if that's the reason, but I think that for the "consistency tests" to pass, you need to have a datalogger installed (not just in the firmware file) but also in `dataloggers.csv`

Also, we need to find a way to reinstate the info around where each datalogger is installed. That is not done in the firmware file, but with a combination of `install/dataloggers.csv` (in which the site code is specified), the info in the `network` folder and possibly the _feature_ file @rumachan was mentioning .

Again, as I said in a couple of comments above, as long as this pr is a placeholder all is fine, but if we want to merge it it might need a bit more work for it to pass.

### comment
`1901` as start date is passing the test, but probably a bit overskilled? maybe use the year since when we purchased those loggers?

**Comment by ricardojohansson** [comment]
Ah, 20 loggers listed here but only 13 of them are included in the dataloggers.csv - I'll update dataloggers.csv now

**Comment by elidana** [comment]
and `assets/dataloggers.csv` as well I think

**Comment by elidana** [comment]
still failing with the duplicate entries message (can't remember what can cause it) but in the meantime

seems that one entry is now missing (possibly in the assets file?) for sn 11300
```
--- FAIL: TestFirmwareHistory_Assets (0.05s)
    --- FAIL: TestFirmwareHistory_Assets/check_for_firmware_assets (0.05s)
        firmware_test.go:87: unable to find firmware receiver asset: CR6 [11300]
```

**Comment by elidana** [comment]
Hi @jacksonshanks and @ricardojohansson , an update for this pull request.
 
Some of the  checks where not passing due to a "silent" bug quickly fixed by @ozym in https://github.com/GeoNet/delta/pull/1268

I'll wait for some under the hood processes to complete and then will review this again now that the silent bug is fixed

**Comment by elidana** [comment]
Hi @ricardojohansson and @rumachan ,

this pull request to add firmware versions of the Campbell dataloggers is now fixed and can be merged.

As you can see from the changes, I have now added leading zeros to all dataloggers' serial numbers (as it is done for other equipment types), so there are a few more files that have been now updated to reflect that.  This is something we need to keep in mind for future entries as well.

Can either of you please review/approve and then this one can be merged?

Thanks!

**Review by ricardojohansson — APPROVED** [review]

**Review by rumachan — APPROVED** [review]
Added leading 0 to CR6 SN with only 4 digits. Good idea as easier to work with.
Carry through the change to install.csv and firmware.csv.

I can't easily check all the dates and SN and firmwares, but the ideas look good.

**Review by elidana — APPROVED** [review]
i auto approve my changes :dancers:

**Pull request merged**